src/quilla/steps/validations/visual_parity.py

```diff
@@ -0,0 +1,308 @@
+        '''
+        Pass-through property to make it easier to call hooks
+        '''
+        return self.ctx.pm.hook
+
+    def _create_report(
+        self,
+        success: bool,
+        msg: str = '',
+        baseline_image_uri: str = '',
+        treatment_image_uri: str = '',
+    ) -> VisualParityReport:
+        return VisualParityReport(
+            success=success,
+            target=self.target,
+            browser_name=self.ctx.driver.name,
+            baseline_id=self.baseline_id,
+            msg=msg,
+            baseline_image_uri=baseline_image_uri,
+            treatment_image_uri=treatment_image_uri
+        )
+
+    def _update_baseline(self):
+
+        baseline_bytes = self.element.screenshot_as_png
+        baseline_image = Image.open(BytesIO(baseline_bytes))
+
+        self.perform_exclusions(baseline_image)
+
+        baseline_bytes = self._get_image_bytes(baseline_image)
+
+        image_uri = self.hook.quilla_store_image(
+            ctx=self.ctx,
+            baseline_id=self.baseline_id,
+            image_bytes=baseline_bytes,
+            image_type=VisualParityImageType.BASELINE
+        )
+
+        if image_uri is None:
+            return self._no_storage_mechanism_report
+
+        if image_uri == '':
+            return self._create_report(
+                success=False,
+                msg='Unable to update the baseline image'
+            )
+
+        return self._create_report(
+            success=True,
+            baseline_image_uri=image_uri,
+            msg='Successfully updated baseline URI'
+        )
+
+    @property
+    def _no_storage_mechanism_report(self):
+        return self._create_report(
+            success=False,
+            msg='No baseline storage mechanism configured',
+        )
+
+    def _get_image_bytes(self, image: Image.Image):
+        buffer = BytesIO()
+        image.save(buffer, format='PNG')
+        buffer.seek(0)
+
+        return buffer.getvalue()
+
+    def perform_exclusions(self, image: Image.Image):
+        '''
+        Using the 'excludeXPaths' parameter, grabs all the necessary elements
+        that should be excluded, and removes them from the screenshot by covering
+        the element position with a black box that is of the same size as the
+        bounding box of that element.
+
+        This mutates the original image, so nothing is returned
+
+        Args:
+            image: The image to perform the exclusions on
+        '''
+
+        exclusion_xpaths = self._parameters.get('excludeXPaths', [])
+
+        for xpath in exclusion_xpaths:
+            self._exclude_element_from_image(image, xpath)
+
+    def _exclude_element_from_image(self, image: Image.Image, exclude_target: str):
+        resolved_exclusion_target = self.ctx.perform_replacements(exclude_target)
+        exclude_element = self.driver.find_element_by_xpath(resolved_exclusion_target)
+
+        if not self._verify_exclude_target_within_bounding_box(exclude_element):
```

Comment:
Even the exclude element is not within the bounding box, it won't have any impact for the final result. Actually I feel it's better to remove this check because sometimes the element to exclude may cover part of the bounding box, for instance, a floating Ad window that partially covers the main page that client may want to remove.

Reply:
My concern is that it could lead to some unexpected behaviour since it attempts to manipulate the image. I can try to do some fancier geometry to determine if the exclude targets are partially within the bounding box, and try to cover those up appropriately, but not performing this check can lead to some weird errors from trying to add components to an image that go outside the image's dimensions